rfcs/999-introduce-release-controller.adoc

```diff
@@ -0,0 +1,126 @@
+* Have better defined semantics for Releases that should be processed at the
+edges--for example, do not process Release any status only updates from the
+lister, since the _Release Controller_ is be the only source of such
+mutations;
+* Enable internal re-enqueuing to force processing of status only
+changes--allowing, for example, idioms like "update status and re-process
+Release" without involving the lister;
+* Have well defined and composable finite state machine to handle act and
+"observe" operations; and probably more.
+
+== Guide level explanation
+
+From Shipper users' point of view, nothing should change.
+
+== Reference level explanation
+
+The current _Schedule Controller_ implementation can be used as the initial
+_Release Controller_ implementation. The _Schedule Controller_ currently
+performs the following tasks on Release objects:
+
+* Fetches and cache the Release's chart;
+* Renders the chart into manifests, and renders those manifests to extract
+relevant information to be annotated in the Release;
+* Creates the Installation, Capacity and Traffic target objects for the Release;
+and
+* Informs other actors the Release has been scheduled by encoding this
+information in the _Scheduled_ condition.
+
+It is expected that those tasks are still performed after the
+implementation of this RFC.
+
+The pipeline composed by the aforementioned tasks should be encoded into
+symmetric *act* and *observe and report* processes.
+
+[[independent-processes]]
+==== Independent *act* and *observe and report* processes
+
+Act processes should _ideally_ not change the Release status, and observe and
+report operations should _ideally_ not solely rely on information present only
+in-memory, uncommitted representation of the Release object.
+
+In other words, act operations should be as independent as possible from their
+observation and reporting counterpart, although we understand that some
+exceptions might exist for this, in the case no other solution is possible.
+
+For example, to perform cluster selection the developer might encode the results
+of this selection as an annotation in the Release in the act operation, and
+read from the lister whether the Release object has the expected annotation
+present and report through the _Scheduled_ condition in the "observe" operation.
+
+Another example is that assessing the Release chart might not necessarily change
+anything in the world, so execution information (the chart is healthy) could be
+encoded directly in the Release conditions right after acting.
+
+==== Composable act operations
+
+The Web Services Middlewares approach can function as a simplified finite state
+machine, where at any point in the execution the processing can be interrupted
+by skipping the next state returning immediately to the caller (this is also
+called link:https://fsharpforfunandprofit.com/rop[Railway Oriented Programming]
+(here is a link:https://github.com/dc0d/rop[Go implementation of ROP] for
+reference).
```

Comment:
Probably we want to be a little careful about adopting super high power abstractions. I think we can probably navigate this just by paying attention to the high level APIs of the "schedule" and "run strategy" modules.